### PR TITLE
remove cloud node_lifecycle_controller's dependencies on k/k

### PR DESCRIPTION
**Pull request by yastij**
Signed-off-by: Yassine TIJANI <ytijani@vmware.com>

**What type of PR is this?**

/kind cleanup
/priority important-soon

**What this PR does / why we need it**: This removes the k/k dependencies of the cloud node_lifecycle_controller

**Which issue(s) this PR fixes**: part of https://github.com/kubernetes/kubernetes/issues/81172

**Special notes for your reviewer**:

/assign @liggitt @andrewsykim 

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: pkg/controller/cloud/node_lifecycle_controller.go [hunk 1]
```diff
@@ -139,7 +138,7 @@ func (c *CloudNodeLifecycleController) MonitorNodes() {
 
 		if status == v1.ConditionTrue {
 			// if taint exist remove taint
-			err = controller.RemoveTaintOffNode(c.kubeClient, node.Name, node, ShutdownTaint)
+			err = cloudnodeutil.RemoveTaintOffNode(c.kubeClient, node.Name, node, ShutdownTaint)
```
Comment (andrewsykim) [inline pkg/controller/cloud/node_lifecycle_controller.go]: Curious where else `RemoveTaintOffNode` and `AddOrUpdateTaintOnNode` are referenced from `pkg/controller`

Comment (yastij) [reply pkg/controller/cloud/node_lifecycle_controller.go]: for `RemoveTaintOffNode` the other in-tree refs are:
```
pkg/controller/util/node
pkg/controller/nodelifecycle/
test/e2e
```

for `AddOrUpdateTaintOnNode ` the in-tree refs are:
```
pkg/controller/util/node
test/e2e
```

File: staging/src/k8s.io/cloud-provider/node/helpers/taints.go [hunk 2]
```diff
@@ -86,6 +86,106 @@ func AddOrUpdateTaintOnNode(c clientset.Interface, nodeName string, taints ...*v
...
+		}
+		if !updated {
+			return nil
+		}
+		return PatchNodeTaints(c, nodeName, oldNode, newNode)
+	})
+}
+
+// TaintExists checks if the given taint exists in list of taints. Returns true if exists false otherwise.
+func TaintExists(taints []v1.Taint, taintToFind *v1.Taint) bool {
```
Comment (andrewsykim) [inline staging/src/k8s.io/cloud-provider/node/helpers/taints.go]: Unexport if there are no external callers

File: staging/src/k8s.io/cloud-provider/node/helpers/taints.go [hunk 3]
```diff
@@ -86,6 +86,106 @@ func AddOrUpdateTaintOnNode(c clientset.Interface, nodeName string, taints ...*v
...
+		if taint.MatchTaint(taintToFind) {
+			return true
+		}
+	}
+	return false
+}
+
+// RemoveTaint tries to remove a taint from annotations list. Returns a new copy of updated Node and true if something was updated
+// false otherwise.
+func RemoveTaint(node *v1.Node, taint *v1.Taint) (*v1.Node, bool, error) {
```
Comment (andrewsykim) [inline staging/src/k8s.io/cloud-provider/node/helpers/taints.go]: Unexport if there are no external callers

File: staging/src/k8s.io/cloud-provider/node/helpers/taints.go [hunk 4]
```diff
@@ -86,6 +86,106 @@ func AddOrUpdateTaintOnNode(c clientset.Interface, nodeName string, taints ...*v
...
+		return newNode, false, nil
+	}
+
+	newTaints, _ := DeleteTaint(nodeTaints, taint)
+	newNode.Spec.Taints = newTaints
+	return newNode, true, nil
+}
+
+// DeleteTaint removes all the taints that have the same key and effect to given taintToDelete.
+func DeleteTaint(taints []v1.Taint, taintToDelete *v1.Taint) ([]v1.Taint, bool) {
```
Comment (andrewsykim) [inline staging/src/k8s.io/cloud-provider/node/helpers/taints.go]: Unexport if there are no external callers

File: staging/src/k8s.io/cloud-provider/node/helpers/taints.go [hunk 5]
```diff
@@ -86,6 +86,106 @@ func AddOrUpdateTaintOnNode(c clientset.Interface, nodeName string, taints ...*v
...
+		if taintToDelete.MatchTaint(&taints[i]) {
+			deleted = true
+			continue
+		}
+		newTaints = append(newTaints, taints[i])
+	}
+	return newTaints, deleted
+}
+
+
```
Comment (andrewsykim) [inline staging/src/k8s.io/cloud-provider/node/helpers/taints.go]: nit: remove extra line

**Review by andrewsykim — COMMENTED** [review]
Overall LGTM, small comments. I think small duplication here is better than trying to update all the existing callers from `pkg/controller`.

**Comment by andrewsykim** [comment]
/approve
/lgtm 

Thanks @yastij

File: staging/src/k8s.io/cloud-provider/node/helpers/taints.go [hunk 6]
```diff
@@ -86,6 +86,105 @@ func AddOrUpdateTaintOnNode(c clientset.Interface, nodeName string, taints ...*v
 	})
 }
 
+// RemoveTaintOffNode is for cleaning up taints temporarily added to node,
+// won't fail if target taint doesn't exist or has been removed.
+// If passed a node it'll check if there's anything to be done, if taint is not present it won't issue
+// any API calls.
+func RemoveTaintOffNode(c clientset.Interface, nodeName string, node *v1.Node, taints ...*v1.Taint) error {
```
Comment (liggitt) [inline staging/src/k8s.io/cloud-provider/node/helpers/taints.go]: if we're duplicating the function, also duplicate the unit test?

**Comment by liggitt** [comment]
one comment on bringing along the test to keep coverage

**Comment by liggitt** [comment]
/hold

feel free to unhold once the test is added

File: staging/src/k8s.io/cloud-provider/node/helpers/taints_test.go [hunk 7]
```diff
@@ -0,0 +1,187 @@
+package helpers
```
Comment (andrewsykim) [inline staging/src/k8s.io/cloud-provider/node/helpers/taints_test.go]: need license header

**Review by andrewsykim — COMMENTED** [review]
Can we keep the test util change in a separate PR and keep this PR focused on the cloud node lifecycle controller changes?

**Pull request closed without merging**